phonetest.py

```diff
@@ -200,18 +201,93 @@ def lookup(cls, phoneid, config_file, chunk):
-              build_url=None):
+              build_url=None, branch=None, revision=None):
 
         logger.debug('PhoneTest.match(tests: %s, test_name: %s, phoneid: %s, '
                      'config_file: %s, job_guid: %s, '
-                     'build_url: %s' % (tests, test_name, phoneid,
-                                        config_file, job_guid,
-                                        build_url))
+                     'build_url: %s, branch: %s, revision: %s' % (
+                      tests, test_name, phoneid, config_file, job_guid,
+                      build_url, branch, revision))
+
         matches = []
         if not tests:
             tests = [PhoneTest.instances[key] for key in PhoneTest.instances.keys()]
 
+        # See if we need to check changeset files so we don't hit the hg
+        # server unnecessarily.
+        has_run_if_changed = False
         for test in tests:
+            if test.run_if_changed:
+                has_run_if_changed = True
+                break
+
+        def get_changeset_files(branch, revision):
+            changeset_files = set()
+            if branch in ['aurora', 'beta']:
+                branch = 'releases/mozilla-' + branch
+            elif branch not in ['mozilla-central']:
+                branch = 'integration/' + branch
+            s = requests.Session()
+            url = 'https://hg.mozilla.org/' + branch + '/json-pushes?changeset=' + revision
+            r = s.get(url,
+                      headers={'Accept': 'application/json',
+                               'User-Agent': 'autophone'})
+
+            if r.status_code != 200:
+                logger.debug('Could not find pushlog for revision %s at %s' %
+                             (revision, url))
+                return set()
+
+            pushlog = r.json()
+            for pushid in pushlog:
+                try:
+                    changesets = pushlog[pushid]['changesets']
+                except:
+                    # We normally see a 'user' in this json blob, more might
+                    # be added in the future.
+                    continue
+                for changeset in changesets:
+                    #TODO: When Bug 1286353 is fixed, use json-rev for this,
+                    #      as this requires retrieving and scanning the whole
+                    #      diff.
+                    url = 'https://hg.mozilla.org/' + branch + '/raw-rev/' + changeset
+                    r = s.get(url,
+                              headers={'User-Agent': 'autophone'})
+                    if r.status_code == 200:
+                        for line in r.text.split('\n'):
+                            if line.find('/dev/null') != -1:
+                                continue
+                            if line.startswith('+++') or line.startswith('---'):
+                                # skip markers, space and leading slash
+                                changeset_files.add(line[6:])
```

Comment:
We really only care about the paths, right? We could take the os.path.dirname and add that to the set.